specs/distros.rst

```diff
@@ -0,0 +1,580 @@
+   but feels too much like Java's setters and getters and so on:
+
+       >>> device.mtu
+       1500
+       # actually changing the value, not the cached value.
+       >>> device.mtu = 1400
+       1400
+
+       vs
+
+       >>> device.mtu()
+       1500
+       >>> device.set_mtu(1400)
+       >>>
+
+ - similar to the previous point, we could have ``is_something`` or
+   ``is_something()``. We must choose one of this variants and use it
+   consistently accross the project.
+
+
+Proposed distro hierarchy
+=========================
+
+Both frameworks have a concept of Distro, each different in its way:
+
+    - cloudinit has a ``distros`` location. There is a ``Distro`` base class,
+      with abstract methods implemented by particular distros.
+
+        Problems:
+
+        * not DRY: many implementations have duplicate code with the base class
+        * not properly encapsulated: distro-specific code executed outside the
+          ``distros`` namespace.
+        * lots of utilities, with low coherence between them.
+
+    - cloudbaseinit has a ``osutils`` location. There is a ``BaseOSUtils``
+      base class, with a WindowsUtils implementation.
+
+       Problems:
+
+           * it's a pure utilities class, leading to low coherence
+             between functions.
+           * it is not a namespace of OS specific functionality.
+             For this, there is also ``utils.windows``.
+
+As seen, both projects lack a namespaced location for all the OS related code.
+
+The following architecture proposal tries to solve this issue, by having one
+namespace for both general utilies related to a distro, as well as distro
+specific code, which doesn't have a counterpart on other distros.
+
+It can have the following advantages:
+
+    * one common place for all distro interaction, with standardized
+      API for each subnamespace and increased coherence.
+
+    * avoids leaky abstractions. Distro specific code goes into ``distros``
+      namespace.
+
+    * eases testability, it is easy to provide a mock with autospeccing
+      for a namespace class, such as Route.
+
+    * Pythonic API, easy to understand and use.
+
+
+The distros location is proposed, with the following structure and attributes:
+
+    - The base class for a distro is found in ``distros.base``.
+
+    - There are sub-namespaces for specific interaction with the OS,
+      such as network, users.
+
+    - More namespaces can be added, if we identify a group of interactions that can
+      be categorized in a namespace.
+
+    - There is a ``general`` namespace, which contains general utilities that can't be moved
+      in another namespace.
+
+    - Each subnamespace has its own abstract base class, which must be implemented
+      by different distros. Code reuse between distros is recommended.
+
+    - Each subnamespace exposes a way to obtain an instance of that namespace for
+      the underlying distro. This can be exposed in namespace.factory.
```

Comment:
Not necessarily a factory, but a function or something to obtain a distro, without caring what the underlying platform is. Something like 'get_network' or 'get_users' should suffice (the design pattern used inside it is irrelevant for the users of the API though)